### PR TITLE
adding h2c graceful shutdown

### PR DESCRIPTION
**Pull request by WeidiDeng**
Fixes [4907](https://github.com/caddyserver/caddy/issues/4907)

### CODE REVIEW

**Comment by mholt** [comment]
Thanks! I'm looking forward to reviewing this.

**Comment by mholt** [comment]
Thank you for this change @WeidiDeng  -- if it's OK with you, I'm going to bump this to after 2.6 (or at least after its initial beta) just so I can stay on a self-imposed schedule :zany_face: 

I will review this though!

File: modules/caddyhttp/h2chandler.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	"net/http"
+	"net/textproto"
+	"sync/atomic"
+	"time"
+
+	"golang.org/x/net/http/httpguts"
+)
+
+// h2chandler is a Handler which counts possible h2c upgrade requests
+type h2chandler struct {
```
Comment (francislavoie) [inline modules/caddyhttp/h2chandler.go:15]: Shouldn't this be uppercase, as a separate word?
```suggestion
type h2cHandler struct {
```

**Comment by mholt** [comment]
Will take another look when it's updated and ready for another review (as discussed in Slack). :100:

**Comment by mholt** [comment]
@WeidiDeng Just tidying things up... I might close this for now if that's alright with you? It's still in a draft state, and seems to have fallen by the wayside.

I'm open to this enhancement, but it does seem rather complex and probably could use some tests as well.

Thank you for having worked on this, and for all you contribute to the project!

**Comment by WeidiDeng** [comment]
Actually the [upstream](https://github.com/golang/net/commit/7191757bc637cf79a7ece0546e33f903bf5e9709) has changed. So this needs to be reworked anyway.

**Pull request closed without merging**